common/src/main/java/io/pravega/common/util/ResourcePool.java

```diff
@@ -0,0 +1,286 @@
+     * it can be reused. 
+     * 
+     * @param t resource to return to the pool.
+     */
+    public void returnResource(T t) {
+        returnResource(t, true);
+    }
+
+    /**
+     * Method to return resource back to the pool. Callers are expected to return the resource to the pool so that
+     * it can be reused. 
+     * If the same resource is returned to the pool multiple times for a single request,
+     * then resource may be given to multiple requests concurrently.
+     *
+     * @param t resource to be returned
+     * @param isValid is resource valid
+     */
+    public void returnResource(T t, boolean isValid) {
+        if (!isValid) {
+            handleInvalid(t);
+        } else {
+            WaitingRequest<T> waiting;
+            boolean toDestroy = false;
+            synchronized (lock) {
+                waiting = waitQueue.poll();
+                if (waiting == null) {
+                    if (!isRunning) {
+                        // The resource will be closed if returned anytime after the shutdown has been initiated.
+                        resourceCount--;
+                        toDestroy = true;
+                    } else {
+                        // as resources are returned to us, we put them in queue to be reused
+                        // if returned resource increases our idle resource count, do not include it
+                        if (idleResources.size() < maxIdle) {
+                            idleResources.offer(t);
+                        } else {
+                            resourceCount--;
+                            toDestroy = true;
+                        }
+                    }
+                }
+            }
+
+            if (waiting != null) {
+                waiting.future.complete(t);
+            }
+
+            if (toDestroy) {
+                if (listener != null) {
+                    listener.notify(Event.Destroyed);
+                }
+                tDestroyer.accept(t);
+            }
+        }
+    }
+
+    private void tryCreateNewResource() {
+        WaitingRequest<T> waiting;
+        synchronized (lock) {
+            if (resourceCount < maxConcurrent) {
+                waiting = waitQueue.poll();
+                if (waiting != null) {
+                    resourceCount++;
+                }
+            } else {
+                waiting = null;
+            }
+        }
+
+        if (waiting != null) {
+            try {
+                tSupplier.get().whenComplete((t, e) -> {
+                    if (e != null) {
+                        waiting.future.completeExceptionally(e);
+                    } else {
+                        if (listener != null) {
+                            listener.notify(Event.Created);
+                        }
+                        waiting.future.complete(t);
+                    }
+                });
+            } catch (Exception e) {
```

Comment:
An assertion error may also be thrown, in which case your future will be left hanging. I suggest catching `Throwable` here. Check out `Futures.completeAfter`